File: leetcode/java/0014. Longest Common Prefix.java
package leetcode.java;

import java.util.HashMap;

// Time complexity: O(S), where S is the total number of characters
// Space complexity: O(S), where S is the total number of characters

class TrieNode {
    public HashMap<Character, TrieNode> node;
    public int count;
    public int total;

    public TrieNode() {
        this.node = new HashMap<>();
        this.count = 0;
        this.total = 0;
    }
}

class Solution {
    public String longestCommonPrefix(String[] strs) {
        TrieNode map = new TrieNode();
        StringBuilder sb = new StringBuilder();

        for (String s : strs) {
            TrieNode temp = map;
            for (char c : s.toCharArray()) {
                if (!temp.node.containsKey(c)) {
                    temp.node.put(c, new TrieNode());
                    temp.count += 1;
                }
                temp.total += 1;
                temp = temp.node.get(c);
            }
        }

        while (map.count == 1 && map.total == strs.length) {
            char key = map.node.entrySet().stream().findFirst().get().getKey();
            sb.append(key);
            map = map.node.get(key);
        }

        return sb.toString();
    }
}

File: leetcode/python/0014. Longest Common Prefix.py
# Time complexity: O(S), where S is the total number of characters
# Space complexity: O(S), where S is the total number of characters

class trie:

    def __init__(self):
        self.storage = {}
        self.next = None

class Solution(object):
    def longestCommonPrefix(self, strs):
        res = []
        store = trie()
        
        for s in strs:
            temp = store
            for char in s:
                if char not in temp.storage:
                    temp.storage[char] = 1
                else:
                    temp.storage[char] += 1
                
                if temp.next is None:
                    temp.next = trie()
                temp = temp.next
        
        while store is not None:
            if len(store.storage) == 1:
                item = store.storage.items()[0]
                if item[1] == len(strs):
                    res.append(item[0])
                    store = store.next
                else:
                    break
            else:
                break
        
        return "".join(res)